File: scripts/exemplo_robo_planar.py
import sys
import os
import numpy as np

# Configurando o path para encontrar a pasta 'src'
caminho_raiz = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
if caminho_raiz not in sys.path:
    sys.path.append(caminho_raiz)

from src.frame import Frame
from src.junta import JuntaRotacional
from src.robo import RoboArticulado

# =======================================================
# 1. CONFIGURAÇÃO DOS FRAMES (Estrutura Física)
# =======================================================
# O Frame 0 é a base. Os outros frames estão deslocados 0.3m no X local do anterior.
# Se fosse 3D, você poderia passar rpy=[np.pi/2, 0, 0] para tombar um eixo.
f1 = Frame(name="F1", origin=[0.0, 0.0, 0.0])
f2 = Frame(name="F2", origin=[0.3, 0.0, 0.0])
f3 = Frame(name="F3", origin=[0.3, 0.0, 0.0])
f_garra = Frame(name="F_Garra", origin=[0.3, 0.0, 0.0]) # A ponta da garra está a 0.3m da J3

# =======================================================
# 2. MODELAGEM DAS JUNTAS (Atuadores)
# =======================================================
# Especificamos explicitamente o eixo 'z' para clareza
j1 = JuntaRotacional(name="J1", frame=f1, axis='z')
j2 = JuntaRotacional(name="J2", frame=f2, axis='z')
j3 = JuntaRotacional(name="J3", frame=f3, axis='z')
j_garra = JuntaRotacional(name="JG", frame=f_garra, axis='z') # Travada (0 rad)

# =======================================================
# 3. MONTAGEM DO ROBÔ
# =======================================================
robo = RoboArticulado(name="Planar_3R", joints=[j1, j2, j3, j_garra])

# =======================================================
# 4. APLICAÇÃO DOS VALORES (Cinemática Direta)
# =======================================================
# Lembra: para o código bater com o "Z para dentro" e "Y para cima" da questão,
# usamos a lógica de inverter o sinal se estivermos usando um sistema destro padrão.
val_j1 = -np.pi/2 # +90 deg Horário
val_j2 =  np.pi/2 # -90 deg Anti-horário
val_j3 = -np.pi/2 # +90 deg Horário
valores_atuacao = [val_j1, val_j2, val_j3, 0.0]

# =======================================================
# 5. RESULTADOS
# =======================================================
print("=== RESULTADOS (MODELO 3D-READY) ===")

posicoes = robo.get_xyz_positions(valores_atuacao)
for i, pos in enumerate(posicoes):
    print(f"Ponto {i}: X={pos[0]:.2f}, Y={pos[1]:.2f}, Z={pos[2]:.2f}")

matrizes = robo.forward_kinematics(valores_atuacao)
m_final = matrizes[-1]

# Calculando a orientação final
phi_rad = np.arctan2(m_final[1, 0], m_final[0, 0])
phi_deg = np.rad2deg(phi_rad)

print("\n=== RESPOSTA FINAL ===")
print(f"x   = {m_final[0, 3]:.1f} m")
print(f"y   = {m_final[1, 3]:.1f} m")
print(f"phi = {phi_deg:.0f}°")

File: scripts/exemplo_robo_planar.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <memory>
#include <iomanip>
#include "frame.hpp"
#include "junta.hpp"
#include "robo.hpp"

using namespace std;

int main() {
    // =======================================================
    // 1. CONFIGURAÇÃO DOS FRAMES (Estrutura Física)
    // =======================================================
    // O Frame 0 é a base. Os outros frames estão deslocados 0.3m no X local do anterior.
    Frame<double> f1("F1", {0.0, 0.0, 0.0}, {0.0, 0.0, 0.0});
    Frame<double> f2("F2", {0.3, 0.0, 0.0}, {0.0, 0.0, 0.0});
    Frame<double> f3("F3", {0.3, 0.0, 0.0}, {0.0, 0.0, 0.0});
    Frame<double> f_garra("F_Garra", {0.3, 0.0, 0.0}, {0.0, 0.0, 0.0});

    // =======================================================
    // 2. MODELAGEM DAS JUNTAS (Atuadores)
    // =======================================================
    // Usamos make_shared para alocar as juntas e permitir polimorfismo seguro
    auto j1 = make_shared<JuntaRotacional<double>>("J1", f1, 'z');
    auto j2 = make_shared<JuntaRotacional<double>>("J2", f2, 'z');
    auto j3 = make_shared<JuntaRotacional<double>>("J3", f3, 'z');
    auto j_garra = make_shared<JuntaRotacional<double>>("JG", f_garra, 'z'); // Travada (0 rad)

    // =======================================================
    // 3. MONTAGEM DO ROBÔ
    // =======================================================
    vector<shared_ptr<Junta<double>>> juntas = {j1, j2, j3, j_garra};
    RoboArticulado<double> robo("Planar_3R", juntas);

    // =======================================================
    // 4. APLICAÇÃO DOS VALORES (Cinemática Direta)
    // =======================================================
    double val_j1 = -M_PI / 2.0; // +90 deg Horário
    double val_j2 =  M_PI / 2.0; // -90 deg Anti-horário
    double val_j3 = -M_PI / 2.0; // +90 deg Horário
    vector<double> valores_atuacao = {val_j1, val_j2, val_j3, 0.0};

    // =======================================================
    // 5. RESULTADOS
    // =======================================================
    cout << "=== RESULTADOS (MODELO 3D-READY) ===" << endl;

    auto posicoes = robo.getXYZPositions(valores_atuacao);
    
    // Configura o cout para exibir floats com precisão fixa
    cout << fixed;

    for (size_t i = 0; i < posicoes.size(); ++i) {
        // Limpando ruídos de ponto flutuante próximos a zero (ex: -0.00)
        double x = abs(posicoes[i][0]) < 1e-9 ? 0.0 : posicoes[i][0];
        double y = abs(posicoes[i][1]) < 1e-9 ? 0.0 : posicoes[i][1];
        double z = abs(posicoes[i][2]) < 1e-9 ? 0.0 : posicoes[i][2];

        cout << "Ponto " << i << ": X=" << setprecision(2) << x 
             << ", Y=" << setprecision(2) << y 
             << ", Z=" << setprecision(2) << z << endl;
    }

    auto matrizes = robo.forwardKinematics(valores_atuacao);
    auto m_final = matrizes.back(); // Pega o último elemento (a matriz da garra)

    // Calculando a orientação final
    // atan2 em C++ funciona exatamente como np.arctan2 (y, x)
    double phi_rad = atan2(m_final[1][0], m_final[0][0]);
    double phi_deg = phi_rad * (180.0 / M_PI);

    // Limpando possíveis ruídos de ponto flutuante da posição final
    double final_x = abs(m_final[0][3]) < 1e-9 ? 0.0 : m_final[0][3];
    double final_y = abs(m_final[1][3]) < 1e-9 ? 0.0 : m_final[1][3];

    cout << "\n=== RESPOSTA FINAL ===" << endl;
    cout << "x   = " << setprecision(1) << final_x << " m" << endl;
    cout << "y   = " << setprecision(1) << final_y << " m" << endl;
    cout << "phi = " << setprecision(0) << phi_deg << "°" << endl;

    return 0;
}